server/src/main/java/org/opensearch/action/bulk/BulkPrimaryExecutionContext.java

```diff
@@ -232,6 +232,7 @@ public void resetForExecutionForRetry() {
         currentItemState = ItemProcessingState.INITIAL;
         requestToExecute = null;
         executionResult = null;
+        retryCounter++;
```

Comment:
Execution of items in a single bulk shard is serialized, so parallel access should not be required, but it looks like the better design choice in my opinion as well to have a thread safe object here.

Reply:
> but it looks like the better design choice in my opinion as well to have a thread safe object here

A lot of operations in this class appear to be not thread safe, presumably because it is not needed due to the serialized access you described. I would only introduce thread safety if it is needed, and then it would need to be done for the whole class and not just this one field.

Reply:
> I would only introduce thread safety if it is needed, and then it would need to be done for the whole class and not just this one field.

Agreed.